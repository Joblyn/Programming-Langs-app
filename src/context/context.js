import { createContext } from "react";

export const RouteContext = createContext(null);

export const NameContext = createContext(null);