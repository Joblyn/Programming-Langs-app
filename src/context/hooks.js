export const useAppContext = (value = null) => {
  let route = value;
  return route;
};