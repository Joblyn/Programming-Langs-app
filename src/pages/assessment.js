import React, { useState, useEffect, createElement } from "react";
import NameIcon from "../assets/images/name_icon.svg";
import ProgressLine from "../components/progress/progressline";
import { LargeButton } from "../components/button/button";
import Question from "../components/question/question";
import { useMediaQuery } from "@mui/material";
import { useParams } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { getQuestions } from "../utilities/redux/actions/questions";
import Loader from "../components/loader/loader";

export default function Assessment() {
  const { language } = useParams();
  const [count, updateCount] = useState(1);
  const [startEnd, setStartEnd] = useState({ start: 0, end: 5 });
  const [data, setData] = useState([]);

  const [last, setLast] = useState(false);
  const remainder = data.length % 5;
  const quotient = parseInt(data.length / 5);
  const matches = useMediaQuery("(min-width:960px)");
  const dispatch = useDispatch();

  const questions = useSelector((state) => state.questions);
  const [questionsState, setQuestionsState] = useState([]);

  const responseStore = useSelector((state) => state.responses);
  const [responses, setResponses] = useState({});

  const [error, setError] = useState({});

  // dispatch action to fetch questions and store at redux store
  useEffect(() => {
    dispatch(getQuestions());
  }, [language, dispatch]);

  // store all language questions from redux store
  useEffect(() => {
    if (questions.isSuccessful) {
      setData(questions.data);
    }
    if (questions.isError) {
      console.error()
      setError({ type: "fetch error", message: error });
    }
  }, [questions, error]);

  // set current questions and responses sets
  useEffect(() => {
    if (startEnd.end === undefined) {
      setQuestionsState(data.slice(startEnd.start));
    } else {
      setQuestionsState(data.slice(startEnd.start, startEnd.end));
    }
  }, [data, startEnd]);

  // check if current questions set is the last and update state
  useEffect(() => {
    if (remainder === 0) {
      if (count === quotient) {
        setLast(true);
      } else {
        setLast(false);
      }
      return;
    }
  }, [count, remainder, quotient]);

  // update responses state
  useEffect(() => {
    setResponses(function () {
      let obj = {};
      questionsState.forEach((question) => {
        let prop = `question_${question.id}`;
        obj = { ...obj, [prop]: responseStore[prop] || "" };
      });
      return obj;
    });
  }, [questionsState, responseStore]);

  const nextQuestionsSet = () => {
    updateCount((count) => count + 1);
    if (remainder) {
      if (count === quotient) {
        setStartEnd(({ end }) => ({ start: end }));
        setLast(true);
        setError({});
      }
    } else {
      setStartEnd(({ end }) => ({ start: end, end: end + 5 }));
      setError({});
    }
  };

  const previousQuestionSet = () => {
    updateCount((count) => count - 1);
    setStartEnd(({ start, end }) => ({ start: start - 5, end: end - 5 }));
    setError({});
  };

  const createCommponent = () => {
    const getProps = (questions, id) => ({
      className: id === questions.length - 1 ? "pb-[173px] md:pb-[100px] question_container" : "question_container",
      key: "question_" + id,
    });
    return questionsState.map((question, id) => {
      let propName = "question_" + question.id;
      return createElement(
        "div",
        getProps(questionsState, id),
        <Question
          propName={propName}
          id={id}
          defaultValue={responseStore[propName]}
          question={question}
          setResponses={setResponses}
        />
      );
    });
  };

  let incompleteResponses = Object.values(responses).includes("");

  return (
    <div className="w-full relative" id="top">
      <div className="flex items-center my-2 w-full">
        <span className="align-middle mr-3">
          <img src={NameIcon} alt="_logo_" className="w-10 h-10" />
        </span>
        <span className="uppercase text-green text-lg font-bold ">
          {`${language} assessment`}
        </span>
      </div>
      {questionsState.length ? (
        <>
          <ProgressLine count={count} totalcount={quotient} />
          {createCommponent()}
          <div className={`${matches ? "px-6 pb-6" : "assessment_footer"}`}>
            {error.type === "incomplete responses" && (
              <p className="text-[red] text-base px-6">{error.message}</p>
            )}
            <div className="flex gap-3">
              {count !== 1 && (
                <LargeButton
                  type="back"
                  innerText={"Back"}
                  action={previousQuestionSet}
                />
              )}
              <LargeButton
                type="next"
                innerText={last ? "Finish" : "Next"}
                action={
                  incompleteResponses
                    ? () =>
                        setError({
                          type: "incomplete responses",
                          message: "Please fill in all fields",
                        })
                    : last
                    ? null
                    : nextQuestionsSet
                }
              />
            </div>
          </div>
        </>
      ) : (
        <Loader />
      )}
    </div>
  );
}
