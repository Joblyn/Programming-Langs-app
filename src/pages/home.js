import React, { useEffect, useState } from "react";
import CustomCard from "../components/card/card";
import Header from "../components/header/header";
import PageLoader from "../components/pageloader/pageloader";
import Languages from "../fixtures/languages";
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';

export default function Home() {
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    setTimeout(() => {
      setLoading(false);
    }, 2000);
  }, []);

  if (loading) {
    return (
      <div className="min-h-screen">
        <PageLoader />
      </div>
    );
  }

  return (
    <div>
      <Header />
      <main className="mt-[6rem] p-2">
        <Box sx={{ flexGrow: 1 }}>
          <Grid
            container
            spacing={{ xs: 2, md: 3 }}
            columns={{ xs: 4, sm: 8, md: 12 }}
          >
            {Languages.map((lang, id) => (
              <Grid item xs={2} sm={4} md={4} key={`lang_${lang}_${id}`}>
                <CustomCard
                  title={lang.title}
                  image={lang.image}
                  alt={lang.alt}
                />
              </Grid>
            ))}
          </Grid>
        </Box>
      </main>
    </div>
  );
}
