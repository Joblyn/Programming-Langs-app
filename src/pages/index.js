export { default as Home } from './home';
export { default as Javascript } from './javascript';
export { default as Python } from './python';
export { default as Java } from "./java";