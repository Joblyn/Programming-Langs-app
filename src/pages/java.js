import React from "react";
import data from "../data/java.json";

export default function Java(){
  return <div>
    
  </div>
}