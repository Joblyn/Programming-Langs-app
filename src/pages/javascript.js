import React from "react";
import { Link } from "react-router-dom";

export default function Javascript() {
  return <div>
    <h1>This is  the javascript page</h1>
    <Link to="/">Home</Link>
  </div>;
}
