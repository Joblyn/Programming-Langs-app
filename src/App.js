import React, { useState, useEffect } from "react";
import "./App.css";
import { Routes, Route } from "react-router-dom";
import * as ROUTES from "./constants/routes";
import { Home, PageNotFound, Assessment, Name } from "./pages";
import Layout from "./containers/layout/layout";
import { AppContext } from "./context/context";
import ReactGA from "react-ga";

// initialize google analytics
let TRACKING_ID;
if (window.location.hostname === "woshipfrontend.interviewblindspots.com") {
  TRACKING_ID = "G-QW0ZP2JZHF";
} else {
  if (window.location.hostname === "damp-shore-08296.herokuapp.com") {
    TRACKING_ID = "G-L3SYMZVBRR";
  } else {
    TRACKING_ID = "G-8WCMDG75N4";
  }
}
ReactGA.initialize(TRACKING_ID);

function App() {
  const [route, setRoute] = useState();
  const [name, setName] = useState();

  useEffect(() => {
    ReactGA.pageview(window.location.pathname + window.location.search);
  }, []);

  return (
    <AppContext.Provider value={{ route, setRoute, setName, name }}>
      <Routes>
        <Route index element={<Home />} />
        <Route path={ROUTES.HOME} element={<Home />} />
        <Route path={ROUTES.ASSESSMENT} exact element={<Layout />}>
          <Route path={ROUTES.NAME} exact element={<Name />} />
          <Route path={ROUTES.LANGUAGE} exact element={<Assessment />} />
        </Route>
        <Route path={ROUTES.NOMATCH} element={<PageNotFound />} />
      </Routes>
    </AppContext.Provider>
  );
}

export default App;
