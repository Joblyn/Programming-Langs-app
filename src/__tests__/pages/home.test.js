import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

import "@testing-library/jest-dom";
import { Home } from "../../pages";
import { act } from "react-dom/test-utils";
import Languages from "../../data/languages";
import { AppContext } from "../../context/context";
import App from "../../App";

// mock usestate setstate from React
const setStateMock = jest.fn();
const useStateMock = (useState) => [useState, setStateMock];
jest.spyOn(React, "useState").mockImplementation(useStateMock);

// mock useEffect from React
jest.spyOn(React, "useEffect").mockImplementation((f) => f());

// mock useNavigate from React
const mockedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedNavigate,
}));

// testing the app at the homepage (/)
describe("<Home />", () => {
  afterEach(cleanup);

  //test the timer loader implementation
  test("rendering the home page", async () => {
    // mock timers
    jest.useFakeTimers();

    const setRoute = () => {};
    const history = createMemoryHistory();
    // render component
    const { container, getByText } = render(
      <Router location={history.location}>
        <AppContext.Provider value={{ setRoute }}>
          <Home />
        </AppContext.Provider>
      </Router>
    );

    // once component is rendered, a loader shows up
    act(() => {
      jest.advanceTimersByTime(100);
    });
    expect(
      container.firstChild.getElementsByClassName("spinner-item").length
    ).toBe(3);
    expect(container.firstChild).toMatchSnapshot();

    // check component after timeout
    act(() => {
      jest.advanceTimersByTime(2000);
    });
    expect(
      container.firstChild.getElementsByClassName("custom-card").length
    ).toBe(Languages.length);
    expect(container.firstChild).toMatchSnapshot();

    jest.useRealTimers();
  });

  // navigates to /asse
  test("navigates to /assessment/name page after a card/language is selected", async () => {
    jest.useFakeTimers();

    const history = createMemoryHistory();
    const { container, getByTestId } = render(
      <Router location={history.location} navigator={history}>
        <App />
      </Router>
    );

    act(() => {
      jest.advanceTimersByTime(2000);
    });

    const user = userEvent.setup();
    // click on a language/card
    const cardButton = container.getElementsByTagName("button")[0];
    expect(cardButton.getAttribute("type")).toBe("button");
    await fireEvent.click(cardButton);
    expect(mockedNavigate).toHaveBeenCalledTimes(1);
    expect(mockedNavigate).toHaveBeenCalledWith("/assessment/name");

    jest.useRealTimers();
  });
});
