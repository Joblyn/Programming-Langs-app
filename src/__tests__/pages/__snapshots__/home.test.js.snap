// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> rendering the home page 1`] = `
<div
  class="min-h-screen"
>
  <div
    class="h-screen flex justify-center items-center"
  >
    <div
      class="spinner"
    >
      <div
        class="spinner-item"
      />
      <div
        class="spinner-item"
      />
      <div
        class="spinner-item"
      />
    </div>
  </div>
</div>
`;

exports[`<Home /> rendering the home page 2`] = `
<div>
  <div
    class="flex flex-row justify-between px-2 md:px-4 py-3 shadow-sm border-green border-b fixed top-0 z-30 w-full bg-white header_grid"
  >
    <div
      class="w-fit"
    >
      <a
        href="/"
      >
        <img
          alt="logo_icon"
          class="w-6 h-6 md:w-10 md:h-10 align-middle"
          src="logo.svg"
        />
      </a>
    </div>
    <div>
      <h1
        class="text-base md:text-xl lg:text-3xl tracking-wide font-bold text-center p-2 uppercase"
      >
        Software Interview Self Assessment
      </h1>
    </div>
    <div
      class="relative flex justify-end"
    >
      <span
        class="toggle"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium text-green css-i4bv87-MuiSvgIcon-root"
          data-testid="FormatAlignJustifyIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 21h18v-2H3v2zm0-4h18v-2H3v2zm0-4h18v-2H3v2zm0-4h18V7H3v2zm0-6v2h18V3H3z"
          />
        </svg>
      </span>
    </div>
  </div>
  <main
    class="mt-[6rem] p-2 max-width-[1250px] flex justify-center"
  >
    <div
      class="MuiBox-root css-i9gxme"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-spacing-md-3 css-1xgjpwf-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-4 MuiGrid-grid-md-4 flex justify-center items-start css-iltpxr-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root w-80 custom-card css-1ri6ub7-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <img
                alt="Shell Scripting"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                height="140"
                src="shell_scripting.jpeg"
                style="height: 200px;"
                width="345"
              />
              <div
                class="MuiCardContent-root bg-white css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-p MuiTypography-gutterBottom text-xl css-anmq6p-MuiTypography-root"
                >
                  Shell Scripting
                </div>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-4 MuiGrid-grid-md-4 flex justify-center items-start css-iltpxr-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root w-80 custom-card css-1ri6ub7-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <img
                alt="Java"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                height="140"
                src="java.jpeg"
                style="height: 200px;"
                width="345"
              />
              <div
                class="MuiCardContent-root bg-white css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-p MuiTypography-gutterBottom text-xl css-anmq6p-MuiTypography-root"
                >
                  Java
                </div>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-4 MuiGrid-grid-md-4 flex justify-center items-start css-iltpxr-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root w-80 custom-card css-1ri6ub7-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <img
                alt="python"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                height="140"
                src="python.png"
                style="height: 200px;"
                width="345"
              />
              <div
                class="MuiCardContent-root bg-white css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-p MuiTypography-gutterBottom text-xl css-anmq6p-MuiTypography-root"
                >
                  Python
                </div>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-4 MuiGrid-grid-md-4 flex justify-center items-start css-iltpxr-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root w-80 custom-card css-1ri6ub7-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <img
                alt="high level design"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                height="140"
                src="hld.png"
                style="height: 200px;"
                width="345"
              />
              <div
                class="MuiCardContent-root bg-white css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-p MuiTypography-gutterBottom text-xl css-anmq6p-MuiTypography-root"
                >
                  High Level Design
                </div>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-4 MuiGrid-grid-md-4 flex justify-center items-start css-iltpxr-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root w-80 custom-card css-1ri6ub7-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <img
                alt="algorithms"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                height="140"
                src="algorithms.png"
                style="height: 200px;"
                width="345"
              />
              <div
                class="MuiCardContent-root bg-white css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-p MuiTypography-gutterBottom text-xl css-anmq6p-MuiTypography-root"
                >
                  Algorithms
                </div>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-4 MuiGrid-grid-md-4 flex justify-center items-start css-iltpxr-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root w-80 custom-card css-1ri6ub7-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <img
                alt="multithreading"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                height="140"
                src="multithreading.png"
                style="height: 200px;"
                width="345"
              />
              <div
                class="MuiCardContent-root bg-white css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-p MuiTypography-gutterBottom text-xl css-anmq6p-MuiTypography-root"
                >
                  Multi Threading
                </div>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-4 MuiGrid-grid-md-4 flex justify-center items-start css-iltpxr-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root w-80 custom-card css-1ri6ub7-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <img
                alt="machine learning"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                height="140"
                src="machinelearning.jpeg"
                style="height: 200px;"
                width="345"
              />
              <div
                class="MuiCardContent-root bg-white css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-p MuiTypography-gutterBottom text-xl css-anmq6p-MuiTypography-root"
                >
                  Machine Learning
                </div>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-4 MuiGrid-grid-md-4 flex justify-center items-start css-iltpxr-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root w-80 custom-card css-1ri6ub7-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <img
                alt="docker & kubernetes"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
                height="140"
                src="docker-kubernetes.png"
                style="height: 200px;"
                width="345"
              />
              <div
                class="MuiCardContent-root bg-white css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-p MuiTypography-gutterBottom text-xl css-anmq6p-MuiTypography-root"
                >
                  Docker & Kubenetes
                </div>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
